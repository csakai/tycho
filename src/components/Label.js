import React from 'react';
import PropTypes from 'prop-types';

class Label extends React.Component {

  static propTypes = {
    position: PropTypes.object,
    text: PropTypes.string.isRequired
  }

  getPosition = (pos) => {
    if (pos) {
      return {
        position: 'absolute',
        top: `${pos.top}px`,
        left: `${pos.left}px`
      };
    }
    return null;
  }
  
  render() {
    return (
      <span style={this.getPosition(this.props.position)}>
        {this.props.text}
      </span>
    );
  }
}

export default Label;
