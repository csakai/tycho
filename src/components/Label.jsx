import React from 'react';
import PropTypes from 'prop-types';

class Label extends React.Component {

  static propTypes = {
    position: PropTypes.object,
    text: PropTypes.string.isRequired,
    active: PropTypes.bool
  }

  getPosition = (pos) => {
    if (pos) {
      return {
        position: 'absolute',
        top: `${pos.top}px`,
        left: `${pos.left}px`
      };
    }
    return null;
  }

  getChildren = (children, active) => {
    if (active) {
      return children;
    }
    return null;
  }
  
  render() {
    return (
      <div>
        <span style={this.getPosition(this.props.position)}>
          {this.props.text}
        </span>
        {this.getChildren(this.props.children, this.props.active)}
      </div>
    );
  }
}

export default Label;
