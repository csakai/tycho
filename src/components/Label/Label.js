import React from 'react';
import PropTypes from 'prop-types';

class Label extends React.Component {

  static propTypes = {
    position: PropTypes.object,
    orbital: PropTypes.object.isRequired
  }

  getPosition = () => {
    const pos = this.props.position;

    if (pos) {
      return {
        position: 'absolute',
        top: `${pos.top}px`,
        left: `${pos.left}px`
      };
    }
    return null;
  }
  
  render() {
    return (
      <span style={this.getPosition()}>
        {this.props.orbital.name}
      </span>
    );
  }
}

export default Label;
