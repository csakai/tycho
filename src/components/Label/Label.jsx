import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';

export default class Label extends React.Component {

  static propTypes = {
    position: PropTypes.object,
    text: PropTypes.string.isRequired,
    onClick: PropTypes.func.isRequired,
    active: PropTypes.bool,
    enabled: PropTypes.bool
  }
  
  render() {
    return (
      <div className={classNames({
        'label--show': this.props.enabled,
        'label--hide': !this.props.enabled,
      })}>
      <span
        className="label__text"
        style={this.props.position}
        onClick={this.props.onClick}>
        {!this.props.active && this.props.text}
      </span>
      {this.props.active && this.props.children}
    </div>
    );
  }
}
