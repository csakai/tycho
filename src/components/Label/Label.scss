.label {
  &--show {
    display: block;
  }

  &--hide {
    display: none;
  }

  &__text {
    position: absolute;
    text-shadow: 1px 1px 1px #c0c0c0;
    opacity: 0.75;
    font-family: Georgia, serif;
    cursor: pointer;

    &:hover {
      opacity: 1;
    }
  }
}
