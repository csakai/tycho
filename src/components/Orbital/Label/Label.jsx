import React from 'react';
import PropTypes from 'prop-types';
import DomLabel from 'three-dom-label';

export default class Label extends React.Component {

    static propTypes = {
        text: PropTypes.string.isRequired,
        id: PropTypes.string.isRequired,
        action: PropTypes.object.isRequired,
        targetId: PropTypes.string
    }

    componentDidMount = () => {
        this.label = this.getLabel();   
        this.refs.group.add(this.label);
    }

    componentWillUnmount = () => this.label.unmount()

    componentWillReceiveProps = (nextProps) => {
        this.maybeUpdateClassName(nextProps);
    }

    /**
     * Updates class name if the active target has changed.
     * 
     * @param {Object} props
     * @param {String} props.id - id of current orbital
     * @param {String} props.targetId - id of active orbital
     */
    maybeUpdateClassName = ({id, targetId}) => {
        if (this.props.targetId !== targetId) {
            const isActive = id === targetId;
            const classSuffix = isActive ? 'active' : 'inactive';

            this.label.setClass(`label label--${classSuffix}`);
        }
    }

    /**
     * Renders a new label, with mouse events bound to it.
     * 
     * @returns {Object3D} label
     */
    getLabel = () => {
        const {id, text, action} = this.props;
        const classPrefix = `label label`;

        return new DomLabel({
            ...this.props,
            events: {
                click: action.setActiveOrbital.bind(this, id, text),
                mouseover: action.addHighlightedOrbital.bind(this, id),
                mouseout: action.removeHighlightedOrbital.bind(this, id)
            },
            className: `${classPrefix}--inactive`
        });
    }

    render() {
        return <group ref="group"></group>;
    }
}