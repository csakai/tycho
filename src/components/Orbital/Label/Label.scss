@import 'globals.scss';

.label {
    &--active {
        animation: glowing-text $transition-speed infinite;
        pointer-events: none;
        font-size: $font-size-medium;
    }

    &--inactive {
        color: $brand-primary;
        cursor: pointer;
        transition: font-size $transition-speed;

        &:hover {
            text-shadow: $text-shadow;
            font-size: $font-size-medium;
        }
    }
}

@keyframes glowing-text {
  0%   { text-shadow: 0 }
  40%  { text-shadow: $text-shadow }
  60%  { text-shadow: $text-shadow }
  100% { text-shadow: 0 }
}