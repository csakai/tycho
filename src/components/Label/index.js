import Label from './Label';
import './Label.css';

export default Label;
