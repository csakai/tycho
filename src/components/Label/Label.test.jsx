import React from 'react';
import toJson from 'enzyme-to-json';
import {shallow} from 'enzyme';
import Label from './Label';

describe('Label Component', () => {
  let component;

  describe('render()', () => {
    describe('when not active', () => {
      beforeEach(() => {
        component = shallow(
          <Label 
            text="foo"
            active={false}
            onClick={jest.fn()}
          />
        );
      });

      it('should render the Label with text and without children if not active', () => {
        expect(toJson(component)).toMatchSnapshot();
      });
    });

    describe('when active', () => {
      beforeEach(() => {
        component = shallow(
          <Label 
            text="foo"
            active={true}
            onClick={jest.fn()}
          />
        );
      });

      it('should render the Label without text and with children if active', () => {
        expect(toJson(component)).toMatchSnapshot();
      });
    });
  });
});
