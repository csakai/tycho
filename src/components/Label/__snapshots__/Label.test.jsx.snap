// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Label Component render() when active should render the Label without text and with children if active 1`] = `
<div
  className="label--hide"
>
  <span
    className="label__text"
    onClick={[Function]}
  />
</div>
`;

exports[`Label Component render() when not active should render the Label with text and without children if not active 1`] = `
<div
  className="label--hide"
>
  <span
    className="label__text"
    onClick={[Function]}
  >
    foo
  </span>
</div>
`;
